java-client/barrage/src/main/java/io/deephaven/client/impl/BarrageSubscriptionImpl.java

```diff
@@ -136,7 +136,7 @@ public void onNext(final BarrageMessage barrageMessage) {
 
                 // if the message was empty, then BaseTable prevents propagating the empty update, and our listener was
                 // not invoked, so let's invoke it ourselves
-                if (numRows == 0) {
+                if (listener != null && numRows == 0) {
```

Comment:
Let's not do this. We're invoking a listener on the wrong thread, potentially without the mutual-exclusion or happens-before boundary we expected. It's racing with the update processing that needs to actually happen for the completion condition to be properly detected. It might even be racing with another notification that is supposed to trigger completion; that is, we might trigger completion early, from the wrong thread, in the case where we get a message with only removes.

 Let's start by reverting the original change.

Then I think we need a deeper change.